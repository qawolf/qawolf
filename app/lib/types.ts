// GraphQL Types
export type ApiKey = {
  created_at: string;
  id: string;
  last_used_at: string | null;
  name: string;
  token: string | null;
  token_end: string;
};

export type AuthenticatedUser = {
  access_token: string;
  user: User;
};

export type DeploymentEnvironment = "all" | "preview" | "production";

export type Env = { [name: string]: string };

export type EnvironmentVariable = {
  created_at: string;
  id: string;
  name: string;
};

export type Group = {
  alert_integration_id: string | null;
  created_at: string;
  deployment_branches: string | null;
  deployment_environment: DeploymentEnvironment | null;
  deployment_integration_id: string | null;
  id: string;
  is_default: boolean;
  is_email_enabled: boolean;
  name: string;
  next_at: string | null;
  repeat_minutes: number | null;
};

export type Integration = {
  github_repo_name: string;
  id: string;
  slack_channel: string | null;
  team_name: string | null;
  type: "github" | "slack";
};

export type Invite = {
  email: string;
  id: string;
  team_id: string;
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

export type RunStatus = "created" | "fail" | "pass";

export type Run = {
  code: string;
  completed_at: string | null;
  current_line: number | null;
  id: string;
  logs_url: string | null;
  started_at: string | null;
  status: RunStatus;
  suite_id: string | null;
  test_id: string;
  video_url: string | null;
};

export type Runner = {
  api_key: string | null;
  ws_url: string | null;
};

export type Suite = {
  created_at: string;
  group_id: string;
  id: string;
  runs: SuiteRun[];
  team_id: string;
};

export type SuiteRun = {
  completed_at: string | null;
  gif_url: string | null;
  id: string;
  is_test_deleted: boolean;
  started_at: string | null;
  status: RunStatus;
  test_id: string;
  test_name: string;
};

export type Team = {
  id: string;
  is_enabled: boolean;
  name: string;
  renewed_at: string | null;
};

export type Test = {
  code: string;
  deleted_at: string | null;
  id: string;
  is_enabled: boolean;
  name: string;
  updated_at: string;
  url: string;
  version: number;
};

export type TestWithSummary = Test & {
  groups: SelectedGroup[];
  summary: {
    gif_url: string | null;
    last_runs: SuiteRun[];
  };
};

export type User = {
  avatar_url: string | null;
  email: string;
  id: string;
  onboarded_at: string | null;
  teams: Team[];
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

// Component Types
export type Action = {
  id: string;
  x: number;
  y: number;
};

export type AuthMode = "logIn" | "signUp";

export type CreateCode = {
  code: string;
  test_id: string | null;
};

export type GroupTests = {
  [testId: string]: string[];
};

export type Modal =
  | "apiKeys"
  | "deployment"
  | "envVariables"
  | "deleteGroup"
  | "deleteTest"
  | "teamSettings";

export type NavigationOption = "code" | "logs";

export type NavigationType = "dark" | "light";

export type SelectedGroup = {
  id: string;
  name: string;
};

export type SelectedIntegration = {
  github_repo_name: string;
  id: string;
};

export type SelectedTest = {
  id: string;
  name: string;
};

export type UserSubscription = {
  email: string;
  wolfName: string;
  wolfVariant: string;
};

export type ValueProp = {
  detail: string;
  message: string;
};

export type Wolf = {
  name: string;
  number: number;
  variant: string;
};

// Runner Types
export type CodeUpdate = {
  code: string;
  generated?: boolean;
  test_id: string;
  version: number;
};

export type Log = {
  message: string;
  severity: "info" | "warning" | "error";
  timestamp: string;
};

export type RunOptions = {
  code: string;
  end_line?: number;
  env: Env;
  restart: boolean;
  run_id?: string;
  start_line?: number;
  test_id?: string;
  version: number;
};

export type RunProgress = {
  code: string;
  completed_at: string | null;
  current_line: number | null;
  run_id?: string;
  start_line?: number;
  status: RunStatus;
};

// State Types
export type ImageGroup = "wolf";

type SignUp = {
  inviteId?: string | null;
  redirectUri?: string | null;
};

type ModalState = {
  group?: SelectedGroup;
  integration?: SelectedIntegration;
  name: Modal | null;
  teamId?: string;
  tests?: SelectedTest[];
};

export type State = {
  dashboardUri: string | null;
  email: string | null;
  error: string | null;
  groupId: string | null;
  image: { wolf: number };
  modal: ModalState;
  run: RunOptions | null;
  signUp: SignUp;
  teamId: string | null;
};
