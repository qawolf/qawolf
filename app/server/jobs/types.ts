export const JOB_TYPES = [
  "checkPending",
  "deleteRunners",
  "deployRunners",
  "orchestrateGroups",
  "orchestrateRunners",
  "restartRunners",
] as const;

export type Job = typeof JOB_TYPES[number];
