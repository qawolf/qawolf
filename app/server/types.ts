import { Transaction } from "knex";

import { Logger } from "./Logger";

// Azure types
export type SaveArtifacts = {
  gifUrl: string | null;
  logsUrl: string;
  videoUrl: string | null;
};

// Context type
export type Context = {
  api_key: string | null;
  ip: string | null;
  logger: Logger;
  teams: Team[] | null;
  user: User | null;
};

// GitHub types
export type GitHubEmail = {
  email: string;
  primary: boolean;
  verified: boolean;
  visibility?: string | null;
};

export type GitHubFields = {
  avatar_url: string;
  email: string;
  github_id: number;
  github_login: string;
  name: string | null;
};

export type GitHubUser = {
  avatar_url: string;
  email: string;
  id: number;
  login: string;
  name: string;
};

// Slack types
export type SlackWebhook = {
  settings_url: string;
  slack_channel: string;
  slack_team: string;
  webhook_url: string;
};

// Model types
export type ApiKey = {
  created_at: string;
  id: string;
  last_used_at: string | null;
  name: string;
  team_id: string;
  token_digest: string;
  token_end: string;
  updated_at?: string;
};

type WolfFields = {
  wolf_name?: string | null;
  wolf_number?: number | null;
  wolf_variant?: string | null;
};

export type CreateUserWithEmail = WolfFields & {
  email: string;
  login_code: string;
};

export type CreateUserWithGitHub = GitHubFields & WolfFields;

export type DeploymentEnvironment = "preview" | "production";

export type EnvironmentVariable = {
  created_at?: string;
  group_id: string;
  id: string;
  is_system?: boolean;
  name: string;
  team_id: string;
  value: string;
  updated_at?: string;
};

export type FormattedVariables = { [name: string]: string };

export type GitHubCommitStatus = {
  completed_at: string | null;
  context: string;
  created_at?: string;
  deployment_url: string;
  github_installation_id: number;
  group_id: string;
  id: string;
  owner: string;
  repo: string;
  sha: string;
  suite_id: string;
  updated_at?: string;
};

export type Invite = {
  accepted_at: string | null;
  created_at?: string;
  creator_id: string;
  email: string;
  expires_at: string;
  id: string;
  team_id: string;
  updated_at?: string;
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

export type Group = {
  alert_integration_id: string | null;
  created_at?: string;
  creator_id: string;
  deleted_at: string | null;
  deployment_branches?: string | null;
  deployment_environment?: DeploymentEnvironment | null;
  deployment_integration_id: string | null;
  id: string;
  is_default: boolean;
  is_email_enabled: boolean;
  name: string;
  next_at: string | null;
  repeat_minutes: number | null;
  team_id: string;
  updated_at?: string;
};

export type GroupTest = {
  group_id: string;
  id: string;
  test_id: string;
};

export type IntegrationType = "github" | "slack";

export type Integration = {
  created_at?: string;
  github_installation_id?: number | null;
  github_repo_id?: number | null;
  github_repo_name?: string | null;
  id: string;
  settings_url: string | null;
  slack_channel?: string | null;
  team_id: string;
  team_name: string | null;
  type: IntegrationType;
  updated_at?: string;
  webhook_url?: string | null;
};

export type ModelOptions = {
  trx?: Transaction;
  logger: Logger;
};

export type RunStatus = "created" | "fail" | "pass";

export type Run = {
  created_at?: string;
  code: string;
  completed_at?: string;
  current_line: number | null;
  id: string;
  started_at?: string;
  status: RunStatus;
  suite_id: string | null;
  test_id: string;
  updated_at?: string;
};

export type Runner = {
  api_key?: string | null;
  created_at: string;
  deleted_at?: string | null;
  deployed_at?: string | null;
  health_checked_at?: string | null;
  id: string;
  location: string;
  ready_at?: string | null;
  restarted_at?: string | null;
  run_id?: string | null;
  session_expires_at?: string | null;
  test_id?: string | null;
  updated_at?: string;
};

export type RunnerLocation = {
  buffer: number;
  latitude: number;
  longitude: number;
  reserved: number;
};

export type RunnerLocations = Record<string, RunnerLocation>;

export type Suite = {
  alert_sent_at?: string;
  created_at: string;
  creator_id: string | null;
  environment_variables: string | null;
  group_id: string;
  id: string;
  team_id: string;
  updated_at?: string;
};

export type SuiteRun = {
  gif_url: string | null;
  id: string;
  is_test_deleted: boolean;
  status: RunStatus;
  test_id: string;
  test_name: string;
};

export type TeamPlan = "free" | "business";

export type Team = {
  created_at?: string;
  deleted_at?: string | null;
  id: string;
  is_enabled: boolean;
  name: string;
  plan: TeamPlan;
  renewed_at: string | null;
  stripe_customer_id: string | null;
  stripe_subscription_id: string | null;
  updated_at?: string;
};

export type TeamUserRole = "admin";

export type TeamUser = {
  id: string;
  role: TeamUserRole;
  team_id: string;
  user_id: string;
};

export type Test = {
  created_at: string;
  creator_id: string;
  code: string;
  deleted_at: string | null;
  id: string;
  is_enabled: boolean;
  name: string;
  runner_locations?: string | null;
  runner_requested_at?: string | null;
  team_id: string;
  url: string;
  updated_at: string;
  version: number;
};

export type TestUpdate = {
  code: string;
  generated?: boolean;
  id: string;
  version: number;
};

export type User = {
  avatar_url: string | null;
  created_at?: string;
  email: string;
  github_id: number | null;
  github_login: string | null;
  id: string;
  is_enabled?: boolean;
  login_code_digest: string | null;
  login_code_expires_at: string | null;
  name: string | null;
  onboarded_at: string | null;
  updated_at?: string;
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

// Mutaton types
export type AuthenticatedUser = {
  access_token: string;
  user: CurrentUser;
};

export type CreateApiKeyMutation = {
  name: string;
  team_id: string;
};

export type CreateGitHubIntegrationsMutation = {
  installation_id: number;
  team_id: string;
};

export type CreateEnvironmentVariableMutation = {
  group_id: string;
  name: string;
  value: string;
};

export type CreateInviteMutation = {
  emails: string[];
  team_id: string;
};

export type CreateSlackIntegrationMutation = {
  group_id: string;
  redirect_uri: string;
  slack_code: string;
};

export type CreateSuiteMutation = {
  group_id: string;
  test_ids?: string[] | null;
};

export type CreateTestMutation = {
  group_id: string | null;
  url: string;
};

export type CreateUrlMutation = {
  redirect_uri: string;
};

export type DeleteGroup = {
  default_group_id: string;
  id: string;
};

export type DeleteTestsMutation = {
  ids: string[];
};

export type InstrumentTestRunMutation = {
  status: RunStatus;
  test_id: string;
};

export type JoinMailingListMutation = {
  email: string;
};

export type RunnerRun = {
  artifacts: SaveArtifacts;
  code: string;
  env: string;
  id: string;
  test_id: string;
  version: number;
};

export type RunTestStatus = {
  message: string;
  pass: boolean;
};

export type SendLoginCodeMutation = {
  email: string;
  invite_id?: string | null;
};

export type SendLoginCode = {
  email: string;
};

export type SignInWithEmailMutation = {
  email: string;
  login_code: string;
};

export type SignInWithGitHubMutation = {
  github_code: string;
  github_state: string;
  invite_id?: string | null;
};

export type UpdateGroupMutation = {
  deployment_branches?: string | null;
  deployment_environment?: DeploymentEnvironment | null;
  deployment_integration_id?: string | null;
  id: string;
  is_email_enabled?: boolean;
  name?: string;
  alert_integration_id?: string | null;
  repeat_minutes?: number | null;
};

export type UpdateGroupTestsMutation = {
  add_group_id: string | null;
  remove_group_id: string | null;
  test_ids: string[];
};

export type UpdateRunMutation = {
  current_line: number | null;
  id: string;
  status: RunStatus;
};

export type UpdateRunnerMutation = {
  id: string;
  is_healthy?: boolean;
  is_ready?: boolean;
};

export type UpdateTeamMutation = {
  id: string;
  name: string;
};

export type UpdateTestMutation = {
  code?: string;
  id: string;
  is_enabled?: boolean;
  name?: string;
  version?: number;
};

export type UpdateUserMutation = {
  onboarded_at: string;
};

export type UseTestSessionMutation = {
  test_id: string;
};

// Query types
export type CurrentUser = User & {
  teams: Team[];
};

export type GroupIdQuery = {
  group_id: string;
};

export type IdQuery = {
  id: string;
};

export type RunResult = Run & {
  logs_url: string | null;
  video_url: string | null;
};

export type RunnerResult = {
  api_key?: string;
  ws_url?: string;
};

export type TeamIdQuery = {
  team_id: string;
};

export type TestQuery = {
  id?: string;
  run_id?: string;
};

export type TestResult = {
  run: RunResult | null;
  test: Test;
};

export type TestSummary = {
  gif_url: string | null;
  last_runs: SuiteRun[];
};

export type Wolf = {
  name: string;
  number: string;
  variant: string;
};
