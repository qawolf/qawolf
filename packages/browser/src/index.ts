// public API
export { BrowserContext, register } from "./ContextManager";
export { Page } from "./page/PageManager";
// TODO move to jest-playwright
export { launch, LaunchResult } from "./launch";

// internal API
export * from "./actions";
export * from "./keyboard";
export { ContextManager } from "./ContextManager";
export { PageManager } from "./page/PageManager";
