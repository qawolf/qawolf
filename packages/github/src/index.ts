import Octokit from "@octokit/rest";

const octokit = new Octokit({
  auth: "d9e1b19607b2b07ed5c6f2e78fa5fb1bd3cabf26"
});

const main = async () => {
  const res = await octokit.checks.create({
    owner: "qawolf",
    repo: "internal",
    name: "QA Wolf",
    head_sha: "9c52a7a6f2090c58bcb23ae066ce31920789f71c"
  });

  console.log("RESPONSE", res);
};

main();
