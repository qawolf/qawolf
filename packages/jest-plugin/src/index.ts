// setup fail-fast since we do not want to continue when a step fails
// hopefully jest-circus implements this behavior, but was not the case when I tested jest 24.9.0
// https://github.com/facebook/jest/issues/2867#issuecomment-476647864
// https://github.com/facebook/jest/issues/6527#issuecomment-506278251

type Callback = (name: string) => void;

const j = jasmine as any;

// require since there are no types
const failFast = require("@qawolf/jasmine-fail-fast");
const jasmineEnv = j.getEnv();
jasmineEnv.addReporter(failFast.init());
