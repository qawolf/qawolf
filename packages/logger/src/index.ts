import { Logger } from "./Logger";

export const browserLogger = new Logger("browser");

export const logger = new Logger("server");
