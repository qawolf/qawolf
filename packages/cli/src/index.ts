#!/usr/bin/env node

import { CONFIG } from "@qawolf/config";
import { logger } from "@qawolf/logger";
import { yellow } from "kleur";
import program from "commander";
import { snakeCase } from "lodash";
import { saveCiTemplate } from "./ci";
import { record } from "./record";
import { test } from "./test";
import { parseUrl } from "./utils";
const { version } = require("../package");

program.version(version);

program.usage("<command> [options]");

let recordCommand = program
  .command("record <url> [name]")
  .description("record a workflow and create a test");

if (CONFIG.development) {
  recordCommand = recordCommand.option(
    "-e, --events",
    "save events (for debugging)"
  );
}

recordCommand.action(async (urlArgument, optionalName, cmd) => {
  const url = parseUrl(urlArgument);
  logger.verbose(`record url "${url.href}"`);

  const name = snakeCase(optionalName || url.hostname!);
  await record(url, name, cmd.events);
});

program
  .command("test [name]")
  .description("run a test")
  .action(async optionalName => {
    await test(optionalName);
  });

program
  .command("azure")
  .description("set up an Azure Pipeline")
  .action(async () => {
    await saveCiTemplate("azure");
  });

program
  .command("circleci")
  .description("set up CircleCI")
  .action(async () => {
    await saveCiTemplate("circleci");
  });

program
  .command("github")
  .description("set up a GitHub Action")
  .action(async () => {
    await saveCiTemplate("github");
  });

program
  .command("gitlab")
  .description("set up GitLab CI/CD")
  .action(async () => {
    await saveCiTemplate("gitlab");
  });

program.arguments("<command>").action(cmd => {
  console.log(yellow(`Invalid command "${cmd}"\n`));
  program.help();
});

program.allowUnknownOption(false);

program.parse(process.argv);

if (!process.argv.slice(2).length) {
  console.log("\n");
  program.outputHelp();
}
