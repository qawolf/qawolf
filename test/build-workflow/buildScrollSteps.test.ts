import { loadFixtures } from '../loadFixtures';
import { buildScrollSteps } from '../../src/build-workflow';
import { ElementEvent } from '../../src/types';

describe('buildScrollSteps', () => {
  let events: ElementEvent[];

  beforeAll(async () => {
    events = (await loadFixtures('infiniteScroll')).events;
  });

  it('builds expected steps', async () => {
    const steps = buildScrollSteps(events);
    expect(steps).toMatchSnapshot();
  });
});
