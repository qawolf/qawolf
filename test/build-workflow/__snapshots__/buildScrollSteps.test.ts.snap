// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildScrollSteps builds expected steps 1`] = `
Array [
  Object {
    "action": "scroll",
    "event": Object {
      "htmlSelector": "<html lang=\\"en\\" qaw_innertext=\\"Item 1 Item 2 Item 3 Item 4 Item 5 Item 6 Item 7 Item 8 Item 9 Item 10 Item 11 Item 12 Item 13 Item 14 Item 15 Item 16 Item 17 Item 18 Item 19 Item 20 Item 21 Item 22 Item 23 Item 24 Item 25 Item 26 Item 27 Item 28 Item 29 Item 30 Item 31 Item 32 Item 33 Item 34 Item 35 Item 36 Item 37 Item 38 Item 39 Item 40 Loading...\\"></html>",
      "isTrusted": true,
      "name": "scroll",
      "page": 0,
      "target": Object {
        "attrs": Object {},
        "children": Array [],
        "name": "html",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583107250144,
      "value": Object {
        "x": 0,
        "y": 125,
      },
    },
    "index": 0,
    "value": Object {
      "x": 0,
      "y": 125,
    },
  },
]
`;
